issues/146 Add optimized handling of incremental statistics for partitioned databases

JonasNjopOlsson opened the pull request:
### Solution overview
Fixes issue https://github.com/olahallengren/sql-server-maintenance-solution/issues/146

Based on code in issue #146 - credits to chf1971 (https://github.com/chf1971)

samot1 committed:
BugFix: mixing mirror and backup to multiple destinations

when you specified both, a mirror and multiple backup destinations, there was a chance, that it would possibly store the same part of a multi file backup on the same mirror instead of alternating because of ORDER BY FilePath.

To fix it a FileNumber column was added and replaced the original by ORDER BY FileNumber

samot1 committed:
Delete dbo.DatabaseBackup.sql

had wrong file name

samot1 committed:
BugFix: mixing mirror and backup to multiple destinations 

when you specified both, a mirror and multiple backup destinations, there was a chance, that it would possibly store the same part of a multi file backup on the same mirror instead of alternating because of ORDER BY FilePath.

To fix it a FileNumber column was added and replaced the original by ORDER BY FileNumber

samot1 committed:
new Parameter @CleanUpStartTime / @CleanUpEndTime

When @CleanUpStartTime and @CleanUpEndTime are specified the delition of old backup files will be only executed, when the procedure was called in this timeframe. 

Reason: xp_deletefile can be very slow, if there are many files in the backup directory (e.g. because LOG backups are taken every minute and cleaned up after 14 days -> 14 * 24 * 60 = 20160 files) and should for this reason only been executed once per night to prevent delays in the main working time.

samot1 committed:
new parameter @CleanupUsesCMD

allows you to use xp_cmdshell instead of xp_delete_file to delete old backup files 

because xp_delete_files could be extremly slow if you do frequent log backups (every 5 min), particularly on SQL 2008 R2 before CU11 (> 2 h runtime)

samot1 committed:
create subdirectory only when it does not exists

create subdirectories for backups / mirror(s) only when they do not exists (otherwise spam in the command log / print output)

samot1 committed:
formating backup command

added spaces and line breaks to the backup command (@CurrentCommand03) for a better readable output

samot1 committed:
Enhanced Mirroring

You can now specify up to 3 mirror destinations (before only one), each with separat cleanup time and cleanup mode (no backup to URL support for this and only 2 mirrors when you are using idera Backup as backup tool)

Instead of using the build in MIRROR parameter (Enterprise feature) you can set @MirrorType to 'COPY' or 'COPY_LATER' to use xp_cmd_shell to invoke ROBOCOPY, XCOPY or COPY to copy the created backup files to the mirror directory / directories. 'COPY' is automatical choosen, if you specify a @MirrorDirectory but have no Enterprise / Developer Edition.

The new parameters @MirrorWhenNotAvailable to @MirrorWhenNotAvailable3 lets you specify, what to do if the mirror destination is not available (temporary network problem, reboot ...) - 'ERROR' (default), 'SKIP' or 'COPY_LATER' (most useful).

When either one of the  @MirrorType or @MirrorWhenNotAvailable parameters is set to 'COPY_LATER' it creates entries with CommandType LIKE 'PENDING%' in the master.CommandLog table. At the next execution it will process this entries (except @SkipPendingMirrorCopies is set to 'Y') and copies / verifies the pending files to the mirrors (if they are available and the @MirrorCleanupTime is not reached). 

The copy-later-feature is particularly useful, if the mirror destination is on a remote site with a slower connection and you do not want to delay your regular every-5-minute-log-backup, just because it takes 10 minutes to transfer this big log backup after your index optimize run. To make full use of it, you should set @MirrorType to 'COPY_LATER', @SkipPendingMirrorCopies to 'Y' and create a second SQL Agent job which calls dbo.DatabaseBackup with @BackupType = 'COPY_PENDING_MIRRORS' (this will skip all the other backup stuff)

samot1 committed:
Faster RESTORE VERIFYONLY

add the parameters @BlockSize, @BufferCount and @MaxTransferSize to the RESTORE VERIFYONLY command too to make it faster (if parameters are specified)

JonasNjopOlsson committed:
issues/146 Add optimized handling of incremental statistics for partitioned databases

Based on code in issue #146 - credits to chf1971 (https://github.com/chf1971)

JonasNjopOlsson commented:
> @JonasNjopOlsson : I see that I used a "CONCAT" at one point in the script. This would break when running on pre-SQL2012. Can you edit this line or even remove it? (it's a "print" at line 1800)

I removed the PRINT statements so the CONCAT was removed as well. You can check the diff of my commit in my pull request for any other version incompatibilities but I think it should be fine.

chf1971 commented:
> I removed the PRINT statements so the CONCAT was removed as well. You can check the diff of my commit in my pull request for any other version incompatibilities but I think it should be fine.

Just checked, looks fine! 
Thanks!
/Charles

rgroenewoudt commented:
Any idea when this might be merged? I'm also running into performance issues with 200 partitions.

JonasNjopOlsson committed:
Merge branch 'upstream/master' into OptimizeForPartitionedDatabases
Conflicts resolved in IndexOptimize.sql

JonasNjopOlsson commented:
The pull request has now been updated to merge cleanly with the main branch.

samot1 committed:
merged with Ola's master version from 2020-12-06

samot1 committed:
merged with Ola's master version from 2020-12-06

samot1 committed:
merged with Ola's master version from 2020-12-06

samot1 committed:
Performance Statistic Update with partitions

Modified the original changes of JonasNjopOlsson (see https://github.com/olahallengren/sql-server-maintenance-solution/pull/296) to make it more in the style of Ola's script

Execution time of a statistic update (with @Execute = 'N') on my 10 TB DWH database with several tables with over 800 partitions (one per day): 
with the original script 3:26:58 
with the modification    0:02:40 (factor 1:77)

samot1 committed:
bugfix: added missing space after UNION ALL

samot1 commented:
commit https://github.com/samot1/sql-server-maintenance-solution/commit/741485d1c949fde17518658877df793dc0d7b76b
and minor bugfix (missing space) at https://github.com/samot1/sql-server-maintenance-solution/commit/92ea7444d21364404a3e7b9fefa458bdf1460177

I made a new merge (to the version of 6th December 2020) and modified Jonas' changes a bit to be more in the style of Ola's script (using a @table variable instead of a #temptable, do no longer need additional variables, because I save the object / statistic id in the table itself)

Performance on my test server on a DWH-Database with several tables with > 800 partitions (@execute was set to 'N'):
with the original script 3:26:58 
with the modification  0:02:40 (factor 1:77)

@olahallengren it would be very nice, if you would implement this - it's a big difference if your nightly job runs 4 h (for all databases) or only 3 min for exact the same task....

JonasNjopOlsson committed:
Merge branch 'master' of https://github.com/olahallengren/sql-server-maintenance-solution into OptimizeForPartitionedDatabases

JonasNjopOlsson committed:
Merge branch 'master' of https://github.com/samot1/sql-server-maintenance-solution into OptimizeForPartitionedDatabases
* Included only changes in IndexOptimize.sql.
* Rewrote the handling of statistics modification data to be in line with the rest of the code, with correct error handling.

JonasNjopOlsson commented:
I have now merged some of the changes from @samot1 into my branch. Thank you for this.

I have also rewritten the code to be even more in line with the @olahallengren solution, outputting statistics data into the new table variable **@tmpStatisticsCache** both for incremental and non-incremental statistics. This helps further streamline the code. This also has the advantage of placing the execution of the per-database SQL statement which retrieves the statistics data within the BEGIN TRY section in this part of the code so that the same error handling applies to all cases.

I have tested the code in the environments accessible to me and it seems to work well in this new revision as well.

JonasNjopOlsson committed:
Merge branch 'master' of https://github.com/olahallengren/sql-server-maintenance-solution into OptimizeForPartitionedDatabases

JonasNjopOlsson committed:
Merge remote-tracking branch 'upstream/master' into OptimizeForPartitionedDatabases

JonasNjopOlsson commented:
This pull request has now been updated to include the latest changes as of today (2022-01-07) from the main branch and to merge cleanly with it.

JonasNjopOlsson committed:
Bug fix in upstream code: When specifying @PartitionLevel = 'N', the index and statistics inventory code still referenced partitions when checking for resumable index operations, causing a runtime error.

JonasNjopOlsson committed:
Merge branch 'olahallengren:master' into OptimizeForPartitionedDatabases

JonasNjopOlsson committed:
Merge branch 'olahallengren:master' into OptimizeForPartitionedDatabases

JonasNjopOlsson committed:
Merge branch 'olahallengren:main' into OptimizeForPartitionedDatabases

JonasNjopOlsson committed:
Merge branch 'olahallengren:main' into OptimizeForPartitionedDatabases

JonasNjopOlsson commented:
This pull request has now been updated to include the latest changes as of today (2025-05-27) from the main branch and to merge cleanly with it.

JonasNjopOlsson commented:
Hi @olahallengren,

Would it be possible to include these changes in a future release of the maintenance solution? We have been running IndexOptimize with these changes in our various, heavily partitioned data warehouse and business intelligence environments for a few years now with good results and no known issues.

Without these proposed changes in place, we can see that the execution times for all index and statistics maintenance is much longer in all our partitioned databases. Looking at the timestamps in the CommandLog table, we can see that the majority of the time spent on running UPDATE STATISTICS is spent not on the command itself but on the time between running the commands, which this code change helps address.

See the following gist to to help produce the statistics: https://gist.github.com/JonasNjopOlsson/7f564cf3dc5077cee9b176344b65eaf5

JonasNjopOlsson committed:
Merge branch 'olahallengren:main' into OptimizeForPartitionedDatabases
